Device Rebuild Feature

behlendorf opened the pull request:
### Motivation and Context

The sequential reconstruction feature adds a more traditional RAID rebuild mechanism to ZFS.  Specifically, it allows for mirror vdevs to be rebuilt in LBA order.  Depending on the pools average block size, overall fragmentation, and the performance characteristics of the devices (SMR) sequential reconstruction can restore redundancy in less time than a traditional healing resilver.  However, it cannot verify block checksums as part of the rebuild.  Therefore a scrub is automatically started when the last active sequential resilver completes.

Sequential reconstruction cannot be supported for RAIDZ configurations but is compatible with the dRAID feature being developed.  This functionality was part of the dRAID PR #10102 but is generically useful and can be merged independently.

### Description

The new `-s` option has been added to the `zpool attach` and `zpool replace` command to request sequential reconstruction be performed instead of healing reconstruction when resilvering.

    zpool attach -s <pool> <existing vdev> <new vdev>
    zpool replace -s <pool> <old vdev> <new vdev>

Regardless of if a sequential or healing resilver was requested the `zpool wait -t resilver` command can be used to block until the resilver completes.

The `zpool status` output was updated to report the progress of a sequential resilver in the same way as a normal healing resilver.  The only difference is the top-level vdev is included in the status output.  This is important because sequential resilvers are handled at the top-level vdev which means multiple sequential resilvers may be in progress as long as they're operating on different top-level vdevs.  Sequential/healing resilvers and scrubs are all still mutually exclusive operations and only one at a time is permitted.

### How Has This Been Tested?

As part of this change the `resilver_restart_*` tests were moved in to the `functional/replacement` directory, and the `replacement` tests were renamed and extended to verify both resilvering and rebuilding.

Performance was tested using a single vdev pool comprised of an 8 TB Seagate Constellation HDD (ST8000NM0075).  The attached and replaced drives were identical HDD devices.  The pools were created with default settings (128K record size,   no compression) and filled with 1TiB of data.

The sustained performance of resilvering and rebuilding is shown in MiB/s for four reference pools with a range of block sizes and fragmentation caused by simulated pool aging.  Additionally, scrub performance has been included for reference.  All testing was performed while the pool was idle.

The key take away from the chart below is that pools containing predominantly small blocks (4k-16k) show approximately a __2x increase in performance__ when rebuilding vs resilvering, thereby effectively halving the replacement time.  Pools with large average block sizes (>=128k) perform the same when rebuilding vs resilvering.
 
![image](https://user-images.githubusercontent.com/148917/82721396-9697f500-9c71-11ea-935f-156bb824facb.png)

Terms:
* "Sequential" pools are freshly created and written to once
* "Random" pools are identical to sequential then aged with a copy/remove loop
* "Large Block" pool filled with N 1GiB files
* "Small Block" pool filled with N copies of the Linux kernel source

Performance tunings to explore:
* ~Up to 16MiB blocks allowed, capping at   1MiB may improve performance~
* ~Survey `zfs_rebuild_queue_limit`~
* ~Survey `zfs_vdev_rebuild_min_active`,   `zfs_vdev_rebuild_min_active`~

__[edit]__ After limiting the maximum rebuild segment size to 1M rebuild and resilver performance in now the same for pools with large blocks.  The bar graph was updated to show the updated performance test results.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tonyhutter commented on cmd/zpool/zpool_main.c:
```diff
@@ -7728,6 +7955,25 @@ status_callback(zpool_handle_t *zhp, void *data)
+		printf_color(ANSI_YELLOW, gettext("One or more devices have "
+		    "been rebuilt, scrubbing the pool is recommended.\n"));
+		printf_color(ANSI_BOLD, gettext("action: "));
+		printf_color(ANSI_YELLOW, gettext("Use 'zpool scrub' to "
```
Maybe include the pool name in the `Use 'zpool scrub'` printf, since you could have multiple pools but only one needs to scrubbed.

behlendorf replied:
The status is printed per pool, so my feeling is there's no much to room for confusion. I tried to model the message after the other "action" messages for consistency.

tonyhutter replied:
Ah, you're right - forgot about that.

tonyhutter commented on tests/zfs-tests/tests/functional/replacement/attach_rebuild.ksh:
```diff
@@ -0,0 +1,175 @@
+	done
+	child_pids=""
+
+        log_must zpool export $TESTPOOL1
+        log_must zpool import -d $TESTDIR $TESTPOOL1
+        log_must zfs umount $TESTPOOL1/$TESTFS1
+        log_must zdb -cdui $TESTPOOL1/$TESTFS1
+        log_must zfs mount $TESTPOOL1/$TESTFS1
```
indents

behlendorf replied:
Fixed.

ahrens commented:
@mmaybee Thanks for raising these points.  Here are my thoughts (including some background info that I'm sure you're already aware of):

> characterize these two algorithms as "verified reconstruction" (aka resilver) and "unverified reconstruction" (aka rebuild)

I like the idea of coming up with more descriptive terms for these - "rebuild" and "resilver" don't actually tell me anything about the difference, they sound like the same thing.  "sequential reconstruction" would be a more positive term than "unverified reconstruction".  And I wonder if there's a term that encapsulates the additional silent-data-corruption-repair that happens with scrub/resilver, in addition to just verification.  Maybe "healing reconstruction" and "sequential reconstruction"?

> I am a bit concerned about how viable (or at least recommendable) it is as an option [to not verify checksums]

I think that for the primary use case of reconstructing 2-way mirrors, doing the checksum-based healing doesn't get you much, because only ditto blocks have any chance of actually being corrected.  That wouldn't be the case for double-parity (and higher) DRAID (or triple-wide mirrors), but see below...

> I think the scrub really is necessary.

The downside of "sequential reconstruction" is that (without a subsequent scrub), it doesn't heal silently-corrupted data.  The most common case where this will be important is with DRAID2.  If one drive has suffered silent data corruption (e.g. the drive silently dropped a write), and a different drive fails, we could reconstruct the failed drive using the bad data from the silently-corrupted drive.  In this case a given stripe would have incorrect data on 2 drives (the incorrectly reconstructed drive and the silently-corrupted drive).  If we read this block (or scrub), we can detect this problem and reconstruct the correct data from the remaining drives.  But if we lose another drive before that happens, then the silently-corrupted block will be lost.

I agree that the scrub is recommended, although personally I wouldn't go so far as to say that it's required for 2-way mirrors.  I would advocate for having ZFS automatically start a scrub after a "sequential reconstruction" completes (even on a 2-way mirror).  If the admin wants to cancel the scrub, I would let them take that risk (even on DRAID2+).

> is there really any realistic example where rebuild+scrub is faster than resilver?

I don't see how it could be faster, since scrub does at least as much work as resilver.

But I think there's value to "sequential reconstruction" because the reconstruction gets you back to having full parity (except for blocks that have been silently corrupted) much faster than resilver.   I'm OK with (slightly) separating the process of whole-disk failure and reconstruction from healing of silently-corrupted blocks.

RAIDZ resilver times especially are so slow that you run a real risk of additional drives failing before the resilver completes.  If the 2nd drive fails after the "sequential reconstruction" but before the scrub completes, you won't lose data (except for silent disk errors).  But if a 2nd drive fails before a "healing reconstruction" completes, you lose everything.  Sun's customers had this problem a decade+ ago, and this problem has only gotten worse as HDD sizes have increased (while IOPS has remained roughly constant).  "sequential reconstruction" is a way for DRAID to fix this problem.

That said, my argument is predicated on "sequential reconstruction" being **much** faster than "healing reconstruction".  The data point in this PR of being 2x faster is good, but I actually expected much better.  Resilvering at ~70MB/s might be good enough.  I'm not sure if 2x is worth the complexity, given the trade-off with checksum verification.  Maybe the "sequential-ish resilver" work has gotten us to the point of resilver being "good enough"?  Changing the indirect block size from 16K to 128K is also an improvement since the "old days".

However, this test case may not be representative of the "worst common case" of resilvering.  For example, since the test case has only 1TB of data, "sequential-ish resilver" may be able to cache and sort a good fraction of all the block pointers, which would not be the case with a more realistically-sized pool (e.g. 100TB+ of data).  For reference, Delphix has customers with >100TB of data with ashift=9, recordsize=8k, compression=on, at ~80% capacity, and ~85% fragmented.  I suspect that this use case would have much worse resilver throughput, even with "sequential-ish resilver".  Showing an improvement closer to 10x (timing "sequential reconstruction" without scrub versus "healing reconstruction") in a realistic workload would be a strong argument for the value of this work.

behlendorf commented:
>  I think we should characterize these two algorithms as "verified reconstruction" (aka rasilver) and "unverified reconstruction" (aka rebuild).

I like it.  Calling this "rebuild" always felt awkward to me as well.  Let me see about updating the documentation, code comments and `zpool status` output to use this updated terminology.  For brevity I think it still makes sense to use "rebuild" internally in the function and structure names since none of that is user visible.

As for the command line option I'm torn but I think there is value in keeping it.  Though I would like to change the new `zpool attach|replace -r` option to `zpool attach|replace -u` for "unverified".  The default behavior is always to resilver so I don't think users would accidentally get in to trouble.

They would need to opt in and the option can be clearly explained in the man page.  In addition, after a unverified rebuild completes (aka rebuild) the `zpool status` output will prominently recommend running `zpool scrub` until a new scrub is complete.  I opted not to automatically kick of a scrub since this would suppress the final rebuild statistics in `zpool status`.  If the specific concern is running multiple unverified rebuilds without scrubbing in between, one option would be to prevent running a new rebuild until the pool has been scrubbed.

>  is there really any realistic example where rebuild+scrub is faster than resilver?

That depends on how you measure.  It can absolutely reduce your exposure time compared to a normal resilver which allows the pool to suffer a new failure without any data loss (most likely).  In the common case we expect the scrub not to discover any blocks which need to be repaired.

ahrens commented:
> I opted not to automatically kick off a scrub since this would suppress the final rebuild statistics in zpool status

I'd like to find a way to both automatically kick off a scrub, and display the final rebuild results in `zpool status`.

> If the specific concern is running multiple unverified rebuilds without scrubbing in between, one option would be to prevent running a new rebuild until the pool has been scrubbed.

I think that would significantly reduce the usefulness of rebuild.  The case where rebuild helps is where a 2nd **or subsequent** drive dies after a rebuild completes, but before a resilver could have completed.  If you have a sequence of 3 drives failing, this new restriction could unnecessarily delay the reconstruction of the 2nd failure, causing the 3rd failure to take out the whole pool.  e.g.:

- 2-way mirror or DRAID1
- first drive fails, is replaced, and is rebuilt
- scrub is running
- 2nd drive fails, we wait for scrub to complete
- we wait so long that a 3rd drive fails, taking out the pool
- if we had started the rebuild when the 2nd drive failed, it would have completed before the 3rd drive failed, thus realizing the benefit of improved rebuild time to save us.

richardelling commented:
Some notes:

- Long ago, in a solar system far away, we tried to come up with a reasonable benchmark for ZFS resilvering. We finally decided it was not possible. We did similar models for other arrays that used traditional RAID, and while they were more predictable, they also have a large variance. In the field, the longest resilver we saw on a busy system took a year to complete (raidz2). Bottom line: don't believe predictions or measurements for resilvers.
- The number of parity "drives" makes the biggest difference in data reliability.  Given the reliability of current enterprise-grade disks, don't worry about multi-disk failures occurring on the order of a week. Yes, we have mathematical models that can be used to evaluate this.
- Most data loss is not due to whole drive failure, but drive replacements are like whole drive failures. There are procedures that can be done to do whole drive replacements for partial failures which will always be faster than rebuild.
- For SMR/CMR drives, the answer is "just say no"
- I know of no customer who would agree to choosing the "unverified" option wrt data. For this reason, I think careful wording is important and realize that the proposed rebuild is almost identical to what other RAID arrays do.

mmaybee commented:
How about introducing a "targeted" scrub? Where we provide a vdev id to the command such that it trims it's traversal to read/verify only block pointers involving the specified drive? This could significantly reduce the time to verify the rebuild. Although this would be a bit complicated for dRAID, as all block pointers reference the same top-level (but we could still figure it out using the offset).

behlendorf commented:
> I'd like to find a way to both automatically kick off a scrub, and display the final rebuild results in zpool status.

The "rebuild" information is still available to `zpool status` when scrubbing.  So it should just be a matter of presenting the output in a reasonable way.  The current PR only shows the last rebuild/resilver/scrub to try and be consistent with that long standing behavior.  I can look in to this.

The original thought was to use a different name for this kind of reconstruction in order to distinguish it from resilvering.  However, if we automatically start a scrub when the rebuild completes perhaps there isn't a meaningful distinction.  This could still be reported as a resilver in `zpool status` and the command line documentation can describe it as requesting "sequential reconstruction".

> The data point in this PR of being 2x faster is good, but I actually expected much better

In order to keep the testing realistic I avoided some of the worst case scenarios.  It's definitely possible to see more than a 2x improvement when the pool is highly fragmented.  But rather than show how bad resilver can be (and without the sequential scan code it can be really bad), I wanted to show the new rebuild code could saturate the vdev despite the fragmentation.

Things get even better with dRAID when there are multiple redundancy groups since when rebuilding to a distributed hot spare it's possible to determine which blocks are degraded and only rebuild those.  For example, if you have a dRAID with 3 redundancy groups you'll only need to rebuild 1/3 of the blocks all the devices can participate.  But digressing in to dRAID specific felt a little out of scope for this PR.

behlendorf commented:
> I'd like to find a way to both automatically kick off a scrub, and display the final rebuild results in zpool status.

Done.  I've update the PR to kick off a scrub when that last active rebuild completes.  While scrubbing both the in progress scrub and final rebuild results are reported by `zpool status`.  The scrub may be cancelled by the user at any time.

> How about introducing a "targeted" scrub?

This is a great idea, but I don't think it's necessary for this PR.  If the scrub takes a little longer to complete because it checks the entire pool that seems fine.  We can always make this improvement at a later date.

>  Maybe "healing reconstruction" and "sequential reconstruction"?

I've been giving this some thought but haven't come up with better names.  Unless I hear a better suggestion I'll update the documentation to use this language in the next day or so.  As for the `zpool status` output I'm leaning towards reporting "sequential reconstruction" and "healing reconstruction" both as resilvering.

behlendorf commented:
@mmaybee @ahrens @richardelling @tonyhutter I've updated this PR with the following noteworthy changes based on the conversation above.  I believe this should address all of the concerns thus far.  Please give it another look.

* When the last active "sequential resilver" completes a scrub is now automatically started.  This addresses the concern that a "sequential resilver" should provide the same assurances by default as a "healing resilver".

* Changed `zpool attach|replace -r` to `zpool attach|replace -s`.  This is no longer referred to as rebuild so the `-r` doesn't make sense, plus `-r` was easy to confuse with "(r)esilver" instead of  "(r)ebuild".  `-s` is now used for (s)equential.

* Updated the language to use "healing reconstruction/resilver" and "sequential reconstruction/resilver" in the user documentation.    They code itself still uses the term rebuild to to make it clear this is different than the dsl scan logic, and the comment at the top of `vdev_rebuild.c` has been updated accordingly with an explanation.

* The `zpool status` command no longer reports "sequential reconstruction" as `rebuilding/rebuilt` in the output but as `resilvering/resilvered` the same as a traditional resilver.  The final "sequential reconstruction" statistics are also included in the output while the scrub is running.  They aren't cleared until a new resilver is run.

* The `zpool wait` command no longer has a "rebuild" target which can be waited on.  Instead waiting on a "resilver" will work for both "healing" and "sequential" resilvers.

mmaybee commented:
Nice, this is an improvement.

> The final "sequential reconstruction" statistics are also included in the output while the scrub is running. They aren't cleared until a new resilver is run.

Is this true of both "healing" and "sequential reconstruction", or only of "sequential"? It would be nice if the statistic for the resilver was somehow annotated (with "verified" perhaps). A "healing reconstruction" would always have this annotation, while a "sequential reconstruction" would get the annotation once a scrub had completed.

behlendorf commented:
> Is this true of both "healing" and "sequential reconstruction", or only of "sequential"? 

Only for sequential resilvers, since healing resilvers and scrubs are both reported with the dsl scan stats.  Sadly we don't have the old healing resilver data when the scrub is running.  We could add something but that causes some backward compatibly issues with the user/kernel interface, so I opted not to change this.

> It would be nice if the statistic for the resilver was somehow annotated (with "verified" perhaps). 

That's an interesting idea.  There's no verified annotation currently, but until the scrub completes after a sequential resilver the `zpool status` output will print a recommendation that the pool needs to be scrubbed.

tonyhutter commented:
Since we're now using the RESILVER_START/FINISH event for both types of resilvers, could we should add a field in the event to say what type of resilver it is.  Specifically, I think you could add it as a "resilver_type=<sequential,healing>" single nvlist to the 3rd arg of spa_event_notify():
```C
spa_event_notify(spa, NULL, NULL,
    scn->scn_phys.scn_min_txg ?
    ESC_ZFS_RESILVER_FINISH : ESC_ZFS_SCRUB_FINISH);
```
I believe that nvlist then gets appended to the normal event nvlist.

We could then key off that value in `cmd/zed/zed.d/resilver_finish-start-scrub.sh` to prevent kicking off a scrub after a sequential resilver.  All this could be added in a follow-on PR, since I don't know if it's worth holding this one up to include it.

behlendorf commented:
@tonyhutter good idea.  I've updated to PR to add the "resilver_type=<sequential,healing>" to the posted event.

mmaybee commented on module/zfs/spa.c:
```diff
@@ -6593,6 +6619,17 @@ spa_vdev_attach(spa_t *spa, uint64_t guid, nvlist_t *nvroot, int replacing)
+	if (rebuild) {
+		/*
+		 * For rebuilds, the only allowable parents are a mirror,
+		 * the root vdev, or a dRAID with fixed width stripes.
```
Since dRAID is not yet present, perhaps this comment should be adjusted?
* For rebuilds, the parent vdev type must support reconstruction using only space maps.
* The only allowable parents are the root vdev or a mirror vdev.

behlendorf replied:
Good catch.  I thought I got all of these in an earlier pass, but clearly I missed one.

mmaybee commented on module/zfs/vdev_rebuild.c:
```diff
@@ -0,0 +1,1106 @@
+ * process will take longer than the healing reconstruction described above.
+ * However, it has that advantage that after the reconstruction first phase
+ * completes redundancy has been restored.  At this point the pool can incur
+ * another device failure without disking data loss.
```
disking -> risking

mmaybee commented on module/zfs/vdev_rebuild.c:
```diff
@@ -0,0 +1,1106 @@
+ *
+ *   - Deferred resilvers using sequential reconstruction are not currently
+ *     supported.  When adding another vdev to an active top-level resilver
+ *     it must be restarted.
```
This is unfortunate. Does this mean that, in a multi-drive failure scenario with 2 spares, if a second device fails before the first rebuild completes, the second drive will not be auto-spared by zed?

behlendorf replied:
In a multi-drive failure scenario the zed will still auto-spare both.  If a second drive in the same top-level vdev fails before the first rebuild completes then resilvering is restarted back at the first space map and both drives are reconstructed.  If the drives are in different top-level vdevs there's no need to restart.  This would be the common case for mirror vdevs, but probably not for dRAID.  I looked in to supporting the deferred resilver feature for sequential rebuilds and it's definitely possible.  But the implementation ended up being uglier than I was comfortable with so I left out this functionality for now.

mmaybee commented on module/zfs/vdev_rebuild.c:
```diff
@@ -0,0 +1,1106 @@
+ *     than healing reconstuction particularly when using using HDDs (or
+ *     especially with SMR devices).  Only allocated capacity is resilvered.
+ *
+ *   - Sequential reconstruction has no awareness of ZFS block boundaries.
```
has no awareness of -> is not constrained by

mmaybee commented on module/zfs/vdev_rebuild.c:
```diff
@@ -0,0 +1,1106 @@
+ *   - Unlike a healing resilver or scrub which are pool wide operations,
+ *     sequential reconstruction is handled by the top-level mirror vdevs.
+ *     This allows for it to be started or canceled on a top-level vdev
+ *     without impacting any other top-level vdevs in the pool.
```
Is it theoretically possible that we could have multiple simultaneous rebuilds happening at once on separate drives (with additional rebuild threads)?

behlendorf replied:
Yes, this is currently supported.  The threads are created and managed per top-level vdev, so we can have multiple rebuilds running for different top-level vdevs.  When a second failure occurs for the same top-level vdev is when we need to restart the rebuild to include that device in the reconstruction.  Much in the same way the original dsl_scan resilvering works, except per top-level vdev instead of pool wide.

jwpoduska commented on module/zfs/vdev_rebuild.c:
```diff
@@ -0,0 +1,1106 @@
+	ASSERT(!vd->vdev_rebuilding);
+
+	dmu_tx_t *tx = dmu_tx_create_dd(spa_get_dsl(spa)->dp_mos_dir);
+	VERIFY0(dmu_tx_assign(tx, TXG_WAIT));
```
Is there a possible deadlock here? It would be very, very unlikely. But, vdev_rebuild_lock is held across the dmu_tx_assign() with TXG_WAIT. So, if that lock is requested at some point in spa_sync(), and this thread is holding it already, it seems like there could be a hang. It may not actually be possible, but I note that the other dmu_tx_assign(), in vdev_rebuild_range() is called w/o the lock being held.

behlendorf replied:
Good thought, but I don't believe so given the way the code is structured.  The `vdev_rebuild_lock` __is__ taken by the sync tasks which are run by `spa_sync()`.  However, no sync tasks can be scheduled until after `vdev_rebuild_initiate()` has run to start a new rebuild for the top-level vdev.  At which point the `dmu_tx_assign()` will have run and this code is unreachable until the rebuild process completes.

jwpoduska commented on module/zfs/vdev.c:
```diff
@@ -4072,17 +4164,37 @@ vdev_stat_update(zio_t *zio, uint64_t psize)
+				vdev_rebuild_phys_t *vrp = &vr->vr_rebuild_phys;
+
+				if (vd->vdev_ops->vdev_op_leaf) {
+					mutex_enter(&tvd->vdev_rebuild_lock);
+					vrp->vrp_bytes_rebuilt += psize;
+					mutex_exit(&tvd->vdev_rebuild_lock);
```
Does it make sense to do this with an atomic_add_64() like the ZIO_FLAG_SCAN_THREAD case above?

behlendorf replied:
Done.

GregorKopka commented:
Interesting feature, thanks for adding. 

One question though:  
> Sequential reconstruction cannot be supported for RAIDZ configurations

I get it for the case where a drive is missing as then it has to be rebuild from the raidz stripes parity with require a metadata walk to locate, but what happens in case the replace is targeting a still online drive? Will `replace -s` apply to the temporary mirror vdev that is created to replace the old drive?

behlendorf commented:
> Will replace -s apply to the temporary mirror vdev that is created to replace the old drive?

That's not supported by the current version.  However, I believe that functionally could be added with some additional development work.
